documentation/adr/0018-algorithm-migration-to-paged-files.md

```diff
@@ -0,0 +1,115 @@
+
+As a consequence:
+ - users are required to develop (and test) separate scripts for each API
+ - users make a conscious choice between “in memory (fast)” and “on disk (slow)” operation.
+
+Mantid have (multiple times) attempted to present seamless in-memory/on-disk operation to their users (for a slightly different use case).
+Each time this either failed for technical reasons or because users did not embrace it,
+and the attempts were abandoned.
+
+One of driving forces for the current PACE project is the support of large datasets,
+which are expected to be increasingly common and of increasing size.
+The new SQW object hides all choice between in-memory and on-disk operation through
+the use of a paged `PixelData` object.
+
+ - provides identical API for small and large datasets on small and large computers
+ - the switch between in-memory and on-disk (paged) operation is automatic as dataset grows
+ - the page size is a configuration parameter that is intended to be user configurable for the machines and datasets they’re using:
+    - a large page size will offer better performance if the whole dataset fits in a single page, 
+    - a moderate page size appears optimal for datasets too large for system RAM
+    - it is hard / impossible / undesirable to calculate optimal page size on a per-machine/per-file basis
+- implementation of a single `cut()` method (or function) that operates on an SQW object (either `this` or one passed as an argument) operating through the public API and agnostic to the size of the datasets.
+
+There are several algorithms that need to be updated to support large datasets.
+Any update must be made in such a way that algorithms that the whole of Horace continues to function as it currently does on "small" datasets
+until a point at which all algorithms are updated.
+During the migration it is also necessary that scripts using a subset of the updated algorithms
+can be created to perform end-to-end testing.
+These should not require code changes or alternate APIs.
+
+
+## Decision
+
+- A new API for SQW object will be created.
+The SQW class will to default to in memory operation (i.e. no paging)
+if a page size is not passed explicitly to the constructor
+until all code has been updated to support paged operation
+
+```matlab
+% unpaged: default page size is 0/inf
+s = sqw(filename)
+
+% paged with specified page size
+s = sqw(filename, page_size)
+```
+- a warning message will be output when operations are performed
+on dataset that require paging to enable a user to understand the performance drop
+- all algorithms will be created with a common API pattern.
+The two APIs will share a common implementation,
+the filename variant creating a (paged) SQW object
+	- `algorithm(filename, outfilename)`
+perform operation on a new SQW object that  will be explicitly created as paged
+	- `algorithm(sqw, [outfilename])`
+perform operation on existing SQW object (may be paged or unpaged)
```

Comment:
The language in this section seems a little confused.

There will be three ways to call an algorithm operating on an SQW object:

1. Using a filename `sqw_obj = algorithm(file_name)`: this will create a paged SQW object from the given file. The algorithm will then operate on the paged SQW object. The paged SQW object will never be visible to the user. A smaller all-in-memory SQW object will be outputted.

2. Using an SQW object `algorithm(sqw_obj)`: the input SQW object will be fully in-memory and so will the output.

3. Using a filename and outputting to a file `algorithm(file_name, out_file_name)`: Equivalent to (1) except the resulting SQW object will be saved to file instead of returned as an object. This provides a method to perform an algorithm that maps a large SQW file to another large SQW file.

Reply:
Good point, I will update the way this is expressed.

You point (2) however is incorrect: if a paged SQW object is passed (e.g. by a developer)  the algorithm will operate on in as paged data. returning a paged object.